src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -100,9 +100,11 @@ public sealed override bool MoveNext()
                     {
                         IEnumerator<TSource> enumerator = enumerable.GetEnumerator();
                         ++_state;
-                        if (enumerator.MoveNext())
+
+                        var canMoveNext = enumerator.MoveNext();
+                        SetEnumerator(enumerator);
```

Comment:
Resolved in 1cdc5bc371a33ed8f2cbb93175a576bd0b342ff7

Reply:
Ah I see, this wouldn't cause any side-effects as SetEnumerator() used to be called after enumerator.MoveNext()? It only really calls Dispose() on the previous enumerator, but could that potentially cause issues?

Reply:
This would be the current code-path
```csharp
var canMoveNext = enumerator.MoveNext(); //MoveNext() called before previous Enumerator's Dispose()

_enumerator?.Dispose(); //SetEnumerator call
_enumerator = enumerator; //SetEnumerator call

if (canMoveNext)
{
    StoreFirst();
    return true;
}
```

What your change would imply:
```csharp
_enumerator?.Dispose(); //SetEnumerator call
_enumerator = enumerator; //SetEnumerator call

if (enumerator.MoveNext()) //MoveNext() called after previous enumerator's Dispose()
{
    StoreFirst();
    return true;
}
```

Now I am not sure if this could ever be an issue where some logic in the previous Enumerators Dispose() Method requires the next Enumerators MoveNext() method to be called before Dispose().

Reply:
It's better to dispose of the first enumerator before movenext'ing on the next one.  We don't need to maintain bug-for-bug compatibility to maintain the other ordering.